### PR TITLE
Login pt fix

### PR DESCRIPTION
**Pull request by CosmicSpectrum**
created `ModuleBE_FailedLoginAttempetDB` that handles all logic for blocking login 
integrated it inside `ModuleBE_AccountDB` in the login flow
created ui component that indicates your user is blocked from logging in and counting down until it's release.

### CODE REVIEW

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	 */
+	public updateFailedLoginAttempt = async (accountId: UniqueId) => {
+		const existingLoginAttempt = await this.query.unique(accountId);
+
+		// if the account doesn't have pre existing document create new.
+		if (!existingLoginAttempt)
+			return this.createNewLoginAttempt(accountId);
+
+		//if the login is already blocked handled blocked login
+		if (existingLoginAttempt.count === this.config.maxLoginAttempts) {
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: should be >=

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+			// if the login is still blocked throw an error
+			if (this.validateLoginBlock(existingLoginAttempt))
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: the correct terminology is assert

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		}
+
+		// update the counter
+		existingLoginAttempt.count++;
+
+		//update the doc
+		const updatedDoc = await this.set.item(existingLoginAttempt);
+
+		// check if login is blocked after the update, if so, handle blocked login
+		if (updatedDoc.count === this.config.maxLoginAttempts)
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: should be >=

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+			return this.handleBlockedLogin(updatedDoc);
+	};
+
+	/**
+	 * After a successful login attempt use this function to validate if user isn't blocked
+	 * or if block time elapsed
+	 * @param accountId The account that successfully logged in
+	 */
+	public validateLoginAttempt = async (accountId: UniqueId) => {
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: this is more like an event, should be on Login successful, which is a dispatch listener that we already have

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+		if (this.validateLoginBlock(loginAttempt))
+			return this.handleBlockedLogin(loginAttempt);
+
+		// clean blocked login if blocked time elapsed
+		return this.clearLoginAttempt(accountId);
+	};
+
+	//######################### Internal Logic #########################
+
+	private validateLoginBlock = (loginAttempt: DB_FailedLoginAttempt) => {
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: isAccountLoginBlocked is a more proper name

File: user-account/src/main/_entity/failed-login-attempt/frontend/ModuleFE_FailedLoginAttempt.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import {ApiStruct_FailedLoginAttempt, DBDef_FailedLoginAttempt, DBProto_FailedLoginAttempt} from '../shared';
+import {DispatcherDef, ThunderDispatcherV3} from '@nu-art/thunderstorm/frontend/core/db-api-gen/types';
+import {ApiDefCaller} from '@nu-art/thunderstorm';
+import {ModuleFE_BaseApi} from '@nu-art/thunderstorm/frontend';
+
+
+export type DispatcherType_FailedLoginAttempt = DispatcherDef<DBProto_FailedLoginAttempt, `__onFailedLoginAttemptsUpdated`>;
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/frontend/ModuleFE_FailedLoginAttempt.ts]: This db entity should not have any FE representation

File: user-account/src/main/frontend/ui/Component_LoginBlocked/Component_LoginBlocked.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		const newTimeLeft = this.state.blockedUntil - currentTimeMillis();
+		if (newTimeLeft <= 0) {
+			clearInterval(this.timerInterval);
+			return this.props.hide();
+		}
+
+		this.setState({timeLeft: newTimeLeft});
+	};
+
+	private renderBlockedTimer = () => {
```
Comment (TacB0sS) [inline user-account/src/main/frontend/ui/Component_LoginBlocked/Component_LoginBlocked.tsx:43]: renderAccountBlockedTimer

**Review by TacB0sS — APPROVED** [review]

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+		// update login timestamp if login is successful and not blocked
+		if (loginAttempt.count < this.config.maxLoginAttempts) {
+			loginAttempt.loginSuccessfulAt = currentTimeMillis();
+			return this.set.item(loginAttempt);
+		}
+
+		// if the login is still blocked throw blocked error
+		if (this.assertLoginBlock(loginAttempt))
+			return this.handleBlockedLogin(loginAttempt);
+	};
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: not sure how this method makes sense..
this suppose to be a onLoginSuccessful, no?

File: user-account/src/main/_entity/failed-login-attempt/shared/api-def.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import {ApiDefResolver} from '@nu-art/thunderstorm';
+
+
+export type ApiStruct_FailedLoginAttempt = {
+	_v1: {}
```
Comment (TacB0sS) [inline user-account/src/main/_entity/failed-login-attempt/shared/api-def.ts]: This file is redundant

**Review by TacB0sS — APPROVED** [review]

File: user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts (hunk 8 above)
Comment (CosmicSpectrum) [reply user-account/src/main/_entity/failed-login-attempt/backend/ModuleBE_FailedLoginAttemptDB.ts]: I call it from the on login successfull dispatcher.. it's just this module implementation.. I just exctrated it to a function in the module that's if...

**Pull request merged**